File: internal/boveda/entrada.go
package boveda

import (
	"encoding/json"
	"reflect"
	"strings"
	"time"
)

// Tipo dice qué clase de cosa guarda una entrada.
type Tipo string

const (
	TipoCredencial Tipo = "credencial"
	TipoNota       Tipo = "nota"
	TipoTarjeta    Tipo = "tarjeta"
	TipoIdentidad  Tipo = "identidad"
)

// Antigua es una contraseña que se sustituyó.
//
// Existe por un caso que pasa más de lo que parece: se cambia la contraseña de
// un servicio, el servicio no se entera —o el cambio no llega a guardarse al
// otro lado— y sin la anterior no se entra. Tiene un precio que hay que decir en
// voz alta y está escrito en `docs/seguridad.md`: **un secreto sustituido sigue
// dentro de la bóveda**, así que se puede borrar a mano y hay un tope.
type Antigua struct {
	Secreto string `json:"secreto"`
	Hasta   string `json:"hasta"` // RFC3339, cuándo dejó de valer
}

// maximoHistorial de contraseñas anteriores por entrada. Diez es de sobra para
// el caso real y evita que una entrada crezca sin freno.
const maximoHistorial = 10

// Entrada es una cosa guardada en la bóveda.
//
// Es una estructura ancha con casi todo opcional, en vez de cuatro tipos
// distintos, y es deliberado: los cuatro comparten título, etiquetas, notas y
// fechas, y la mitad del valor de una bóveda está en buscar por encima de todos
// a la vez. Cuatro tipos obligarían a repetir esa búsqueda cuatro veces.
type Entrada struct {
	// ID es inmutable y se pone al crear. Hace falta desde el primer día: sin
	// identidad estable no hay sincronización ni compartir, y ponerlo después
	// sería una migración de datos de verdad.
	ID     string `json:"id"`
	Tipo   Tipo   `json:"tipo"`
	Titulo string `json:"titulo"`

	Notas     string   `json:"notas,omitempty"`
	Etiquetas []string `json:"etiquetas,omitempty"`
	Carpeta   string   `json:"carpeta,omitempty"`

	Creada   string `json:"creada"`
	Cambiada string `json:"cambiada"`

	// Papelera es borrado suave. Barato ahora e imprescindible para sincronizar
	// después: sin él, «borrada aquí» y «nunca existió allí» son indistinguibles.
	Papelera  bool   `json:"papelera,omitempty"`
	BorradaEn string `json:"borradaEn,omitempty"`

	// Credencial.
	Usuario string `json:"usuario,omitempty"`
	Secreto string `json:"secreto,omitempty"`
	// Sitios en plural desde el principio: el autorrelleno va a necesitar varios
	// dominios por entrada, y la misma cuenta se usa en más de un dominio más a
	// menudo de lo que parece.
	Sitios []string `json:"sitios,omitempty"`
	// TOTP es el secreto en base32 tal como lo da el servicio, no el código de
	// seis dígitos. El código se calcula al enseñarlo.
	TOTP      string    `json:"totp,omitempty"`
	Historial []Antigua `json:"historial,omitempty"`

	// Tarjeta.
	Titular      string `json:"titular,omitempty"`
	Numero       string `json:"numero,omitempty"`
	Caduca       string `json:"caduca,omitempty"`
	Verificacion string `json:"verificacion,omitempty"`

	// Identidad.
	NombreCompleto  string `json:"nombreCompleto,omitempty"`
	Documento       string `json:"documento,omitempty"`
	NumeroDocumento string `json:"numeroDocumento,omitempty"`

	// Extra guarda **los campos que esta versión de Esfinge no entiende**.
	//
	// Es lo más subestimado de todo el formato. En cuanto haya dos Esfinges de
	// versiones distintas sobre la misma bóveda —el escenario de la
	// sincronización, y también el de un cliente que no actualiza— la vieja
	// borraría en silencio lo que escribió la nueva. Conservarlos cuesta el
	// Unmarshal/Marshal a medida de abajo; no conservarlos cuesta datos.
	Extra map[string]json.RawMessage `json:"-"`
}

// CambiarSecreto pone una contraseña nueva y guarda la anterior.
func (e *Entrada) CambiarSecreto(nuevo string, ahora time.Time) {
	if e.Secreto != "" && e.Secreto != nuevo {
		e.Historial = append([]Antigua{{
			Secreto: e.Secreto,
			Hasta:   ahora.UTC().Format(time.RFC3339),
		}}, e.Historial...)
		if len(e.Historial) > maximoHistorial {
			e.Historial = e.Historial[:maximoHistorial]
		}
	}
	e.Secreto = nuevo
	e.Cambiada = ahora.UTC().Format(time.RFC3339)
}

// Coincide dice si la entrada encaja con lo que se busca.
//
// Busca por título, usuario, sitios, etiquetas y carpeta. **Nunca por el
// secreto**: teclear en un buscador algo que se compara contra contraseñas es
// una forma silenciosa de averiguarlas a base de probar.
func (e Entrada) Coincide(q string) bool {
	q = strings.ToLower(strings.TrimSpace(q))
	if q == "" {
		return true
	}
	campos := append([]string{e.Titulo, e.Usuario, e.Carpeta, e.NombreCompleto, e.Titular},
		append(e.Sitios, e.Etiquetas...)...)
	for _, c := range campos {
		if strings.Contains(strings.ToLower(c), q) {
			return true
		}
	}
	return false
}

// SinSecretos devuelve la entrada con lo sensible fuera.
//
// **Es lo que cruza el puente hacia la ventana en la lista.** La contraseña, el
// TOTP y el historial se piden de uno en uno y solo cuando se van a ver: eso
// hace más por que un secreto no acabe en un volcado de memoria que todo el
// borrado de búferes junto, porque en cuanto algo cruza a la interfaz vive en el
// montón del webview, fuera de nuestro alcance.
func (e Entrada) SinSecretos() Entrada {
	e.vaciarLoSensible()
	return e
}

// vaciarLoSensible quita de la entrada todo lo que hay que proteger.
//
// **Está en un solo sitio a propósito**, y lo usan dos caminos que parecen
// distintos y no lo son: lo que viaja en la lista hacia la ventana y lo que
// queda en la papelera al borrar. Con dos listas separadas ya pasó lo que tenía
// que pasar: la de borrar solo quitaba la contraseña, el TOTP y el historial, y
// **una nota segura borrada se quedaba entera dentro del fichero**, igual que el
// número de una tarjeta. Añadir un campo sensible y acordarse de dos sitios es
// una defensa que dura hasta la siguiente prisa.
func (e *Entrada) vaciarLoSensible() {
	e.Secreto = ""
	e.TOTP = ""
	e.Historial = nil
	e.Verificacion = ""
	e.Numero = ""
	e.NumeroDocumento = ""
	e.Notas = ""
}

// ---------------------------------------------------------------------------
// Conservación de campos desconocidos.
//
// La lista de claves conocidas se saca por reflexión de las etiquetas de la
// propia estructura, para que añadir un campo no obligue a acordarse de
// actualizar una lista aparte. Es justo la clase de lista que se queda atrás.

var clavesConocidas = func() map[string]bool {
	m := map[string]bool{}
	t := reflect.TypeOf(Entrada{})
	for i := 0; i < t.NumField(); i++ {
		etiqueta := t.Field(i).Tag.Get("json")
		nombre, _, _ := strings.Cut(etiqueta, ",")
		if nombre != "" && nombre != "-" {
			m[nombre] = true
		}
	}
	return m
}()

type entradaCruda Entrada // sin los métodos, para no entrar en bucle

func (e *Entrada) UnmarshalJSON(b []byte) error {
	var cruda entradaCruda
	if err := json.Unmarshal(b, &cruda); err != nil {
		return err
	}
	*e = Entrada(cruda)

	var todo map[string]json.RawMessage
	if err := json.Unmarshal(b, &todo); err != nil {
		return err
	}
	for k := range todo {
		if clavesConocidas[k] {
			delete(todo, k)
		}
	}
	if len(todo) > 0 {
		e.Extra = todo
	}
	return nil
}

func (e Entrada) MarshalJSON() ([]byte, error) {
	b, err := json.Marshal(entradaCruda(e))
	if err != nil {
		return nil, err
	}
	if len(e.Extra) == 0 {
		return b, nil
	}

	var todo map[string]json.RawMessage
	if err := json.Unmarshal(b, &todo); err != nil {
		return nil, err
	}
	for k, v := range e.Extra {
		// Lo conocido manda: un campo que esta versión entiende no se pisa con
		// una copia vieja que venga de Extra.
		if !clavesConocidas[k] {
			todo[k] = v
		}
	}
	return json.Marshal(todo)
}

File: navegador/src/nucleo/entrada.ts
/**
 * Una entrada de la bóveda, igual que `Entrada` en `internal/boveda/entrada.go`.
 *
 * Dos reglas de Go que aquí hay que cumplir a mano, porque de ellas depende que
 * la forma canónica salga igual en los dos lados:
 *
 * - **Los vacíos no se escriben** (`omitempty`), salvo `id`, `tipo`, `titulo`,
 *   `creada` y `cambiada`, que van siempre. Y en el historial, `secreto` y `hasta`
 *   también van siempre, y nada más: Go no conserva lo desconocido ahí dentro.
 * - **Lo desconocido se conserva** en `extra`, tal cual, y lo conocido manda: una
 *   copia vieja de un campo que se entiende no pisa el de verdad.
 */

import { canonico, type ValorJSON } from "./canon";

export type Tipo = "credencial" | "nota" | "tarjeta" | "identidad";

export type Antigua = { secreto: string; hasta: string };

export type Entrada = {
  id: string;
  tipo: Tipo | string;
  titulo: string;
  notas?: string;
  etiquetas?: string[];
  carpeta?: string;
  creada: string;
  cambiada: string;
  revision?: number;
  papelera?: boolean;
  borradaEn?: string;
  usuario?: string;
  secreto?: string;
  sitios?: string[];
  totp?: string;
  historial?: Antigua[];
  titular?: string;
  numero?: string;
  caduca?: string;
  verificacion?: string;
  nombreCompleto?: string;
  documento?: string;
  numeroDocumento?: string;
  /** Los campos que esta versión no conoce. */
  extra?: Record<string, ValorJSON>;
};

/** El orden y la clase de cada campo conocido, que es el orden de la estructura de Go. */
const CAMPOS: [keyof Entrada, "siempre" | "texto" | "lista" | "numero" | "si" | "historial"][] = [
  ["id", "siempre"],
  ["tipo", "siempre"],
  ["titulo", "siempre"],
  ["notas", "texto"],
  ["etiquetas", "lista"],
  ["carpeta", "texto"],
  ["creada", "siempre"],
  ["cambiada", "siempre"],
  ["revision", "numero"],
  ["papelera", "si"],
  ["borradaEn", "texto"],
  ["usuario", "texto"],
  ["secreto", "texto"],
  ["sitios", "lista"],
  ["totp", "texto"],
  ["historial", "historial"],
  ["titular", "texto"],
  ["numero", "texto"],
  ["caduca", "texto"],
  ["verificacion", "texto"],
  ["nombreCompleto", "texto"],
  ["documento", "texto"],
  ["numeroDocumento", "texto"],
];
const CONOCIDOS = new Set<string>(CAMPOS.map(([k]) => k as string));

export const MAXIMO_HISTORIAL = 10;

class ErrorDeForma extends Error {}

function texto(v: unknown): string {
  if (v === null || v === undefined) return "";
  if (typeof v !== "string") throw new ErrorDeForma("Una entrada tiene un campo de texto que no es texto");
  return v;
}

/** Lee una entrada del JSON del cuerpo, como `json.Unmarshal` en Go. */
export function entradaDesde(crudo: unknown): Entrada {
  if (crudo === null || typeof crudo !== "object" || Array.isArray(crudo)) {
    throw new ErrorDeForma("Una entrada no es un objeto");
  }
  const o = crudo as Record<string, unknown>;
  const e: Entrada = { id: "", tipo: "", titulo: "", creada: "", cambiada: "" };
  for (const [k, clase] of CAMPOS) {
    const v = o[k as string];
    switch (clase) {
      case "siempre":
      case "texto": {
        const t = texto(v);
        if (clase === "siempre" || t !== "") (e as Record<string, unknown>)[k] = t;
        break;
      }
      case "lista": {
        if (v === null || v === undefined) break;
        if (!Array.isArray(v)) throw new ErrorDeForma("Una entrada tiene una lista que no es una lista");
        const l = v.map(texto);
        if (l.length > 0) (e as Record<string, unknown>)[k] = l;
        break;
      }
      case "numero": {
        if (v === null || v === undefined) break;
        if (typeof v !== "number" || !Number.isInteger(v)) throw new ErrorDeForma("La revisión no es un número entero");
        if (v !== 0) e.revision = v;
        break;
      }
      case "si": {
        if (v === null || v === undefined) break;
        if (typeof v !== "boolean") throw new ErrorDeForma("La papelera no es sí o no");
        if (v) e.papelera = true;
        break;
      }
      case "historial": {
        if (v === null || v === undefined) break;
        if (!Array.isArray(v)) throw new ErrorDeForma("El historial no es una lista");
        const h = v.map((a) => {
          const x = (a ?? {}) as Record<string, unknown>;
          return { secreto: texto(x.secreto), hasta: texto(x.hasta) };
        });
        if (h.length > 0) e.historial = h;
        break;
      }
    }
  }
  const extra: Record<string, ValorJSON> = {};
  for (const [k, v] of Object.entries(o)) {
    if (!CONOCIDOS.has(k)) extra[k] = v as ValorJSON;
  }
  if (Object.keys(extra).length > 0) e.extra = extra;
  return e;
}

/** La entrada como objeto JSON, con los vacíos fuera y lo desconocido dentro. */
export function entradaAJSON(e: Entrada): Record<string, ValorJSON> {
  const out: Record<string, ValorJSON> = {};
  for (const [k, clase] of CAMPOS) {
    const v = e[k];
    switch (clase) {
      case "siempre":
        out[k] = (v as string | undefined) ?? "";
        break;
      case "texto":
        if (v) out[k] = v as string;
        break;
      case "lista":
        if (Array.isArray(v) && v.length > 0) out[k] = v as string[];
        break;
      case "numero":
        if (v) out[k] = v as number;
        break;
      case "si":
        if (v) out[k] = true;
        break;
      case "historial":
        if (Array.isArray(v) && v.length > 0) {
          out[k] = (v as Antigua[]).map((a) => ({ secreto: a.secreto ?? "", hasta: a.hasta ?? "" }));
        }
        break;
    }
  }
  for (const [k, v] of Object.entries(e.extra ?? {})) {
    if (!CONOCIDOS.has(k)) out[k] = v;
  }
  return out;
}

/** La forma canónica de una entrada (docs/formato-boveda.md). */
export function canonEntrada(e: Entrada): string {
  return canonico(entradaAJSON(e));
}

/** Una copia que no comparte listas ni mapas con el original. */
export function copiar(e: Entrada): Entrada {
  return entradaDesde(JSON.parse(JSON.stringify(entradaAJSON(e))));
}

/** La entrada con lo sensible fuera: lo que cruza a la ventana o al panel en una lista. */
export function sinSecretos(e: Entrada): Entrada {
  const c = copiar(e);
  delete c.secreto;
  delete c.totp;
  delete c.historial;
  delete c.verificacion;
  delete c.numero;
  delete c.numeroDocumento;
  delete c.notas;
  return c;
}

/** Busca por título, usuario, carpeta, nombre, titular, sitios y etiquetas. **Nunca por el secreto.** */
export function coincide(e: Entrada, q: string): boolean {
  const b = q.trim().toLowerCase();
  if (!b) return true;
  const campos = [e.titulo, e.usuario, e.carpeta, e.nombreCompleto, e.titular, ...(e.sitios ?? []), ...(e.etiquetas ?? [])];
  return campos.some((c) => (c ?? "").toLowerCase().includes(b));
}

/** Pone una contraseña nueva y guarda la anterior, como `CambiarSecreto` en Go. */
export function cambiarSecreto(e: Entrada, nuevo: string, ahora: Date): void {
  const cuando = rfc3339(ahora);
  if (e.secreto && e.secreto !== nuevo) {
    e.historial = [{ secreto: e.secreto, hasta: cuando }, ...(e.historial ?? [])].slice(0, MAXIMO_HISTORIAL);
  }
  e.secreto = nuevo;
  e.cambiada = cuando;
}

/** Fecha en RFC3339 a segundos y en UTC, que es como las escribe Go. */
export function rfc3339(d: Date): string {
  return d.toISOString().replace(/\.\d{3}Z$/, "Z");
}
